lib/core/src/sync/mod.rs

```diff
@@ -1 +1,153 @@
+    RecordType, SyncData, SyncOutgoingDetails, SyncState,
+};
+use self::model::{DecryptedRecord, SyncSettings};
+
+pub(crate) mod client;
 pub(crate) mod model;
+
+pub(crate) struct SyncService {
+    remote_url: String,
+    persister: Arc<Persister>,
+    signer: Arc<Box<dyn Signer>>,
+    client: Box<dyn SyncerClient>,
+}
+
+impl SyncService {
+    pub(crate) fn new(
+        remote_url: String,
+        persister: Arc<Persister>,
+        signer: Arc<Box<dyn Signer>>,
+        client: Box<dyn SyncerClient>,
+    ) -> Self {
+        Self {
+            remote_url,
+            persister,
+            signer,
+            client,
+        }
+    }
+
+    async fn pull_changes(&self) -> Result<Vec<Record>> {
+        let latest_revision = self
+            .persister
+            .get_sync_settings()?
+            .map(|settings| settings.latest_revision)
+            .unwrap_or(0);
+
+        let req = ListChangesRequest::new(latest_revision, self.signer.clone())?;
+        Ok(self.client.list_changes(req).await?.changes)
+    }
+
+    fn apply_record(&self, decrypted_record: &DecryptedRecord) -> Result<()> {
+        match decrypted_record.data {
+            SyncData::Chain(chain_data) => self.persister.insert_chain_swap(chain_data.into())?,
+            SyncData::Send(send_data) => self.persister.insert_send_swap(send_data.into())?,
+            SyncData::Receive(receive_data) => {
+                self.persister.insert_receive_swap(receive_data.into())?
+            }
+        }
+        Ok(())
+    }
+
+    fn fetch_sync_data(&self, data_id: &str, record_type: RecordType) -> Result<SyncData> {
+        let data = match record_type {
+            RecordType::Receive => {
+                let receive_data: ReceiveSyncData = self
+                    .persister
+                    .fetch_receive_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Receive swap {data_id}"))?
+                    .into();
+                SyncData::Receive(receive_data)
+            }
+            RecordType::Send => {
+                let send_data: SendSyncData = self
+                    .persister
+                    .fetch_send_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Send swap {data_id}"))?
+                    .into();
+                SyncData::Send(send_data)
+            }
+            RecordType::Chain => {
+                let chain_data: ChainSyncData = self
+                    .persister
+                    .fetch_chain_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Chain swap {data_id}"))?
+                    .into();
+                SyncData::Chain(chain_data)
+            }
+        };
+        Ok(data)
+    }
+
+    async fn apply_incoming(&self) -> Result<()> {
+        let incoming_records = self.pull_changes().await?;
+        let latest_revision = incoming_records.last().map(|record| record.revision);
+
+        for new_record in incoming_records {
```

Comment:
Some points here to think about:
1. Perhaps it is better to first save all incoming records and update the last revision in the settings? Then we can query and iterate one by one because there could be also previous records in the db that we want to handle.
2. Currently if there is an error in handling one record the whole process fails, not sure it is a good idea.
3. If in the middle of the sync there are local updates to a record it seems that we ends up in unexpected state and loose data. Let's think on how to change the process to maintain data integrity.

Reply:
> Then we can query and iterate one by one because there could be also previous records in the db that we want to handle

Makes sense, with the new approach there's no update to > SCHEMA_VERSION records. Could be added to the `clean`, but maybe it makes more sense there.

> Currently if there is an error in handling one record the whole process fails, not sure it is a good idea.

Agreed, I was planning to change each unwrap into an if let Err() once we get to the thread, but it makes more sense doing it now

> If in the middle of the sync there are local updates to a record it seems that we ends up in unexpected state and loose data

Yes, if in between the merge and the push there's an update, it won't be synced outwards. What's worse is that after the push we should clean the sync_outgoing, so we won't even mark it as updatable any longer. Perhaps the idea is, in the clean, to check if the updated_fields_json has changed. If it has, that means the record has received new updates, so we do not remove it from sync_outgoing just yet (we remove old fields of course, just keeping the actual dirty ones). This ensures that, on the next iteration, changes will be pushed. Let me know what you think

Reply:
I would say even more than that. If the fields json has changed, ignore the sync for this record, it will be synced on a later iteration. Partial updates are tricky.